File: tests/algorithms/x/Java/graphs/scc_kosaraju.java
public class Main {

    static long[] dfs(long u, long[][] graph, boolean[] visit, long[] stack) {
        if (((Boolean)(visit[(int)(u)]))) {
            return stack;
        }
visit[(int)(u)] = true;
        for (long v : graph[(int)(u)]) {
            stack = ((long[])(dfs(v, ((long[][])(graph)), ((boolean[])(visit)), ((long[])(stack)))));
        }
        stack = ((long[])(java.util.stream.LongStream.concat(java.util.Arrays.stream(stack), java.util.stream.LongStream.of(u)).toArray()));
        return stack;
    }

    static long[] dfs2(long u, long[][] reversed_graph, boolean[] visit, long[] component) {
        if (((Boolean)(visit[(int)(u)]))) {
            return component;
        }
visit[(int)(u)] = true;
        component = ((long[])(java.util.stream.LongStream.concat(java.util.Arrays.stream(component), java.util.stream.LongStream.of(u)).toArray()));
        for (long v : reversed_graph[(int)(u)]) {
            component = ((long[])(dfs2(v, ((long[][])(reversed_graph)), ((boolean[])(visit)), ((long[])(component)))));
        }
        return component;
    }

    static long[][] kosaraju(long[][] graph) {
        long n = graph.length;
        long[][] reversed_graph_1 = ((long[][])(new long[][]{}));
        long i_1 = 0L;
        while (i_1 < n) {
            reversed_graph_1 = ((long[][])(java.util.stream.Stream.concat(java.util.Arrays.stream(reversed_graph_1), java.util.stream.Stream.of(new long[]{})).toArray(long[][]::new)));
            i_1 = i_1 + 1;
        }
        i_1 = 0;
        while (i_1 < n) {
            for (long v : graph[(int)(i_1)]) {
reversed_graph_1[(int)(v)] = ((long[])(java.util.stream.LongStream.concat(java.util.Arrays.stream(reversed_graph_1[(int)(v)]), java.util.stream.LongStream.of(i_1)).toArray()));
            }
            i_1 = i_1 + 1;
        }
        boolean[] visit_1 = ((boolean[])(new boolean[]{}));
        i_1 = 0;
        while (i_1 < n) {
            visit_1 = ((boolean[])(appendBool(visit_1, false)));
            i_1 = i_1 + 1;
        }
        long[] stack_1 = ((long[])(new long[]{}));
        i_1 = 0;
        while (i_1 < n) {
            if (((Boolean)(visit_1[(int)(i_1)])) == false) {
                stack_1 = ((long[])(dfs(i_1, ((long[][])(graph)), ((boolean[])(visit_1)), ((long[])(stack_1)))));
            }
            i_1 = i_1 + 1;
        }
        i_1 = 0;
        while (i_1 < n) {
visit_1[(int)(i_1)] = false;
            i_1 = i_1 + 1;
        }
        long[][] scc_1 = ((long[][])(new long[][]{}));
        long idx_1 = stack_1.length - 1;
        while (idx_1 >= 0) {
            long node_1 = stack_1[(int)(idx_1)];
            if (((Boolean)(visit_1[(int)(node_1)])) == false) {
                long[] component_1 = ((long[])(new long[]{}));
                component_1 = ((long[])(dfs2(node_1, ((long[][])(reversed_graph_1)), ((boolean[])(visit_1)), ((long[])(component_1)))));
                scc_1 = ((long[][])(java.util.stream.Stream.concat(java.util.Arrays.stream(scc_1), java.util.stream.Stream.of(component_1)).toArray(long[][]::new)));
            }
            idx_1 = idx_1 - 1;
        }
        return scc_1;
    }

    static void main() {
        long[][] graph = ((long[][])(new long[][]{new long[]{1}, new long[]{2}, new long[]{0, 3}, new long[]{4}, new long[]{}}));
        long[][] comps_1 = ((long[][])(kosaraju(((long[][])(graph)))));
        long i_3 = 0L;
        while (i_3 < comps_1.length) {
            System.out.println(java.util.Arrays.toString(comps_1[(int)(i_3)]));
            i_3 = i_3 + 1;
        }
    }
    public static void main(String[] args) {
        main();
    }

    static boolean[] appendBool(boolean[] arr, boolean v) {
        boolean[] out = java.util.Arrays.copyOf(arr, arr.length + 1);
        out[arr.length] = v;
        return out;
    }
}

File: tests/algorithms/x/PHP/graphs/scc_kosaraju.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
$__start_mem = memory_get_usage();
$__start = _now();
  function dfs($u, $graph, &$visit, $stack) {
  if ($visit[$u]) {
  return $stack;
}
  $visit[$u] = true;
  foreach ($graph[$u] as $v) {
  $stack = dfs($v, $graph, $visit, $stack);
};
  $stack = _append($stack, $u);
  return $stack;
};
  function dfs2($u, $reversed_graph, &$visit, $component) {
  if ($visit[$u]) {
  return $component;
}
  $visit[$u] = true;
  $component = _append($component, $u);
  foreach ($reversed_graph[$u] as $v) {
  $component = dfs2($v, $reversed_graph, $visit, $component);
};
  return $component;
};
  function kosaraju($graph) {
  $n = count($graph);
  $reversed_graph = [];
  $i = 0;
  while ($i < $n) {
  $reversed_graph = _append($reversed_graph, []);
  $i = $i + 1;
};
  $i = 0;
  while ($i < $n) {
  foreach ($graph[$i] as $v) {
  $reversed_graph[$v] = _append($reversed_graph[$v], $i);
};
  $i = $i + 1;
};
  $visit = [];
  $i = 0;
  while ($i < $n) {
  $visit = _append($visit, false);
  $i = $i + 1;
};
  $stack = [];
  $i = 0;
  while ($i < $n) {
  if ($visit[$i] == false) {
  $stack = dfs($i, $graph, $visit, $stack);
}
  $i = $i + 1;
};
  $i = 0;
  while ($i < $n) {
  $visit[$i] = false;
  $i = $i + 1;
};
  $scc = [];
  $idx = count($stack) - 1;
  while ($idx >= 0) {
  $node = $stack[$idx];
  if ($visit[$node] == false) {
  $component = [];
  $component = dfs2($node, $reversed_graph, $visit, $component);
  $scc = _append($scc, $component);
}
  $idx = $idx - 1;
};
  return $scc;
};
  function main() {
  $graph = [[1], [2], [0, 3], [4], []];
  $comps = kosaraju($graph);
  $i = 0;
  while ($i < count($comps)) {
  echo str_replace('false', 'False', str_replace('true', 'True', str_replace('"', '\'', str_replace(':', ': ', str_replace(',', ', ', json_encode($comps[$i], 1344)))))), PHP_EOL;
  $i = $i + 1;
};
};
  main();
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
